### ISSUE TITLE
Change return type of valueOf to prevent assignment

### ISSUE THREAD

**Issue by mrcljx** [js-temporal/temporal-polyfill#178]
Currently the return type of `valueOf` is `never`. This is correct
as it will never return a value but always throw.

It certainly helps in situations like this:

```
const v = instant.valueOf()
console.log(v)
```

where TypeScript will complain about unreachable code after the
`valueOf` invocation. But a problem is that `never` is always
compatible with other types, and means that the following snippet
will not raise an error at compile time (while it will at runtime).

```
type MyInterface = { valueOf: () => number }
const value: MyInterface = Temporal.Now.instant()
```

A few options of what one could do to make TypeScript point out
the issue here:

1. Mark `valueOf` as `private`,
2. Make `valueOf` optional,
3. Make `valueOf` return `never | SOMETHING` with `unknown`, `undefined`, or a `unique symbol`.

I added an exploration of these options in a playground.

https://www.typescriptlang.org/play?ssl=36&ssc=1&pln=1&pc=1#code/GYVwdgxgLglg9mABFApgZygCgG4C5EDei2AhgDYgoDyw+mAlIgLwB8iYIAtgEYoBOiAL6MCAKESIICNHDIoAdGTgBzHPNIVqwBogDUiACwAmegG5Rg0aJhhUfYCQgpEAYRB8+KWwEkwGEraE4sTklDR0jKzsKNj8FlZSflCS7p62+G4eXlC+-oFMhCGa4Yg6UURQABZ8cADu0fUAoh5wfDqCQqKoGJgQqdmMAPSD7HCI-DV8VjZ2Dk6IAGokfDABUACCQRIaYcAA-BHMbGAxcZaiiRiIJPhLK2ubBUQ7WoflyNV1DYjNk+2d3SwJHoVgAJigIGRls5LskAHKnPj4cAwACOlEQaAAnjxZOZrLZ+HNnHdVrYAEJbIq7N7HRGIAA+yCxAAcUHBgIgEbEpudYYhuLdlmSoJSntTXqVImwKp96icmi02owOpZAZhuCCLlC0GhFsK1i4qSyVqRUBKaAx8CceVSJFUavKUIq-iCJJY+dJkhBmN9SYaGF10FgICCgA

**Comment by shicks** [comment]
I [updated][1] your playground example to also call `valueOf()` directly.  Interestly, I couldn't reproduce the "unreachable code" error with the current version (can you link to an example of this?).

[1]: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFApgZygCgG4C5EDei2AhgDYgoDyw+mAlIgLwB8iYIAtgEYoBOiAL6MCAKESIICNHDIoAdGTgBzHPNIVqwBogDUiACwAmegG5Rg0QBMUEMiT4pJ0qIgByKbP3zgYAR0pENABPHllzURgwVD5gEggnAGEQPkdogEkwDBJowmJySho6RlZ2T34hSOj+OITEADUHGByoAEE8jULgAH5i5jYwcoFLKJjap0a+ZuiAIQ6CrT7Swa8BAB9kYIAHFDhgdyHKuxI0NAamlsS8ranSVHzNGgZ8FYqiKAALPjgAdzK-gCiqTgfAYpiElWstnsjmcWVcEBSaSg+GSqRQGXhLXMqAwmER6OiZkQAHoSew4Ih+N8+KIpPDJEiAIzMRmEqDqBZPMx06SyBRKVQEvhMnlQ46w+kYRAkfCTaZtHHoLAkHlS1wkFlMGWcx7aNV8uSKFSYTVimwSpzqxDcOUXWZKvHcA0M7ham26rpg3lZfnG1Ru83QhxWlySO1TS6OrAQF3SiDuiCerTeqV+wX40XmIA

Here are my thoughts on the three options:

1. Making `valueOf` private seems promising, assuming it's reasonable to be a class.  Obviously it's not possible for an interface.
2. Making `valueOf` optional seems somewhat unsatisfying.  One possible consideration is that if subclasses are possible, then they can override an optional method with a required method (though it would still have to return `never` for covariance).
3. Note that `never|x` is the same as just `x`, since `never` is the union identity.  So I don't think this is a good option - it's also the only option of the three that permits direct calls to `valueOf` without any complaint at all.

**Comment by mrcljx** [comment]
> I couldn't reproduce the "unreachable code" error with the current version (can you link to an example of this?).

You are right. I must have imagined this working, but [it actually does not](https://github.com/microsoft/TypeScript/issues/37998).

> Note that `never|x` is the same as just `x`

You are right there well. I meant to actually remove that suggestion. `unknown` or `unique symbol` would be the remaining options to at least prevent assignment and still making the value somewhat unusable.
